File: Sources/VoxLocalCore/Models/DictationState.swift
import Foundation

/// Lifecycle states of one dictation session.
public enum DictationState: String, CaseIterable, Sendable, Equatable {
    case idle
    case preparing
    case recording
    case stopping
    case transcribing
    case refining
    case inserting
    case completed
    case cancelled
    case error
}

public struct InvalidTransitionError: Error, Equatable, CustomStringConvertible {
    public let from: DictationState
    public let to: DictationState
    public var description: String { "invalid dictation transition \(from.rawValue) → \(to.rawValue)" }
}

/// Explicit state machine for the dictation lifecycle. Rejects transitions
/// that are not part of the documented graph, which prevents overlapping
/// sessions and out-of-order pipeline steps.
public struct DictationStateMachine: Sendable {
    public private(set) var state: DictationState

    public static let allowedTransitions: [DictationState: Set<DictationState>] = [
        .idle: [.preparing],
        .preparing: [.recording, .cancelled, .error],
        .recording: [.stopping, .cancelled, .error],
        .stopping: [.transcribing, .cancelled, .error],
        .transcribing: [.refining, .inserting, .cancelled, .error],
        .refining: [.inserting, .cancelled, .error],
        .inserting: [.completed, .cancelled, .error],
        .completed: [.idle],
        .cancelled: [.idle],
        .error: [.idle],
    ]

    public init(state: DictationState = .idle) {
        self.state = state
    }

    /// True while a session is in flight (a new session must not start).
    public var isActive: Bool {
        state != .idle
    }

    /// True while the session can still be cancelled by the user.
    public var isCancellable: Bool {
        switch state {
        case .preparing, .recording, .stopping, .transcribing, .refining, .inserting:
            return true
        default:
            return false
        }
    }

    public func canTransition(to next: DictationState) -> Bool {
        Self.allowedTransitions[state]?.contains(next) ?? false
    }

    public mutating func transition(to next: DictationState) throws {
        guard canTransition(to: next) else {
            throw InvalidTransitionError(from: state, to: next)
        }
        state = next
    }
}

File: windows/src/VoxLocal.App/Core/Models/DictationState.cs
namespace VoxLocal.Core.Models;

/// <summary>Lifecycle states of one dictation session.</summary>
public enum DictationState
{
    Idle,
    Preparing,
    Recording,
    Stopping,
    Transcribing,
    Refining,
    Inserting,
    Completed,
    Cancelled,
    Error,
}

public sealed class InvalidTransitionException : Exception
{
    public DictationState From { get; }
    public DictationState To { get; }

    public InvalidTransitionException(DictationState from, DictationState to)
        : base($"invalid dictation transition {from} → {to}")
    {
        From = from;
        To = to;
    }
}

/// <summary>
/// Explicit state machine for the dictation lifecycle. Rejects transitions
/// that are not part of the documented graph, which prevents overlapping
/// sessions and out-of-order pipeline steps.
/// </summary>
public sealed class DictationStateMachine
{
    public DictationState State { get; private set; }

    public static readonly IReadOnlyDictionary<DictationState, IReadOnlySet<DictationState>> AllowedTransitions =
        new Dictionary<DictationState, IReadOnlySet<DictationState>>
        {
            [DictationState.Idle] = new HashSet<DictationState> { DictationState.Preparing },
            [DictationState.Preparing] = new HashSet<DictationState> { DictationState.Recording, DictationState.Cancelled, DictationState.Error },
            [DictationState.Recording] = new HashSet<DictationState> { DictationState.Stopping, DictationState.Cancelled, DictationState.Error },
            [DictationState.Stopping] = new HashSet<DictationState> { DictationState.Transcribing, DictationState.Cancelled, DictationState.Error },
            [DictationState.Transcribing] = new HashSet<DictationState> { DictationState.Refining, DictationState.Inserting, DictationState.Cancelled, DictationState.Error },
            [DictationState.Refining] = new HashSet<DictationState> { DictationState.Inserting, DictationState.Cancelled, DictationState.Error },
            [DictationState.Inserting] = new HashSet<DictationState> { DictationState.Completed, DictationState.Cancelled, DictationState.Error },
            [DictationState.Completed] = new HashSet<DictationState> { DictationState.Idle },
            [DictationState.Cancelled] = new HashSet<DictationState> { DictationState.Idle },
            [DictationState.Error] = new HashSet<DictationState> { DictationState.Idle },
        };

    public DictationStateMachine(DictationState state = DictationState.Idle)
    {
        State = state;
    }

    /// <summary>True while a session is in flight (a new session must not start).</summary>
    public bool IsActive => State != DictationState.Idle;

    /// <summary>True while the session can still be cancelled by the user.</summary>
    public bool IsCancellable => State is DictationState.Preparing
        or DictationState.Recording
        or DictationState.Stopping
        or DictationState.Transcribing
        or DictationState.Refining
        or DictationState.Inserting;

    public bool CanTransition(DictationState next) =>
        AllowedTransitions.TryGetValue(State, out var allowed) && allowed.Contains(next);

    public void Transition(DictationState next)
    {
        if (!CanTransition(next))
        {
            throw new InvalidTransitionException(State, next);
        }
        State = next;
    }
}
